gguf-py: reduce peak RAM during convert by streaming dtype casts

igloo58 opened the pull request:
**Summary**

Fixes #15623. When converting very large HF models to GGUF, we sometimes OOM even with `--use-temp-file`. The main culprit is a NumPy `astype(...)` on huge tensors, which materializes a full-sized temporary array before writing.

This PR teaches the GGUF lazy writer to **stream pure dtype casts** to disk in fixed-size chunks, capping peak RAM during the write.

**What’s changed**

- In `gguf-py/gguf/lazy.py`, detect lazy nodes that are *only* a dtype cast (`astype`) and, in `tofile()`, write them in chunks rather than materializing the whole array first.
- New env knob: `GGUF_CAST_CHUNK_MB` (default **256**) to control the chunk size.
  - Example (macOS/Linux):
    ```bash
    GGUF_CAST_CHUNK_MB=128 python -u convert_hf_to_gguf.py <in_dir> --outfile out.gguf --outtype f16 --use-temp-file
    ```
  - Example (Windows PowerShell):
    ```powershell
    $env:GGUF_CAST_CHUNK_MB="128"
    python .\convert_hf_to_gguf.py <in_dir> --outfile out.gguf --outtype f16 --use-temp-file
    ```

**Behavioral notes**

- No GGUF format changes; output is identical to the eager path.
- Only affects NumPy `astype`-only nodes; complex ops still use the existing path.
- If a tensor is already in the target dtype, no extra work is done.

**Motivation / background**

Even with `--use-temp-file`, the pipeline could still OOM on 100B+ models because `astype` creates a large transient array prior to streaming. The new path avoids that by converting and writing in slices.

**Local results (M1 Pro, macOS 15 / Python 3.10):**

- TinyLlama-1.1B → **F16**: wrote 2.20 GB
  - `maximum resident set size`: ~2.51 GB  
  - `peak memory footprint`: ~0.73 GB
- Qwen2.5-7B-Instruct → **F16**: wrote 15.2 GB
  - `maximum resident set size`: ~4.44 GB  
  - `peak memory footprint`: ~2.62 GB

These numbers reflect bounded peaks during cast/write; without streaming, peaks can jump much higher on large layers.

**Limitations / follow-ups**

- Only streams pure dtype casts; future work could stream other large, simple transforms.
- Defaults to 256 MB per chunk; reviewers can suggest a different default if preferred.

**Testing**

- Converted multiple models end-to-end (`--use-temp-file`) and verified GGUF loads and runs as expected.
- Manually varied `GGUF_CAST_CHUNK_MB` (128/256/512) to confirm peak memory scales with chunk size.

Thanks!

igloo58 committed:
gguf-py: stream dtype casts during GGUF write; add GGUF_CAST_CHUNK_MB env knob; remove duplicate imports

compilade commented:
The max memory usage of the lazy writer *should* be around the size of the biggest tensor (because of `astype` and other transformations which can't necessarily be streamed).

Note that `--use-temp-file` isn't really meant to

compilade commented:
Chunking `astype` will only reduce memory usage in very simple cases where the tensors are not transformed before their type is changed, which excludes the case of MoE models which require stacking the experts tensors (because that doesn't directly involve `astype`). The stacked tensors are still materialized before being type-casted and/or written.

The max memory usage of the lazy writer *should* already be around (at most 3×) the size of the biggest tensor (because of BF16, and/or transformations which can't necessarily be streamed).

The peak extra memory used by `astype` (when *not* chunking writes), as a fraction of the input tensor size is likely is around:

|From/To| F32 | F16 | BF16 |
|-|-|-|-|
|F32 | 0 | 0.5× | 0.5× \* |
|F16| 2× | 0 | 1× \* |
|BF16| 2× \*| 3× \* | 3× \* |

When chunking `astype`, the peak extra memory usage would tend towards (in ideal conditions; still relative to the input tensor size):

|From/To| F32 | F16 | BF16 |
|-|-|-|-|
|F32 | 0 | 0 | 0.5× \* |
|F16| 0 | 0 | 1× \* |
|BF16| 2× \*| 2× \* | 3× \* |

(\* BF16 doesn't use `astype`, and round-trips to F32 before `.numpy()` is called on the Torch tensor. Conversion back to BF16 is handled by `gguf.quantize()`. Chunking `astype` wouldn't reduce its memory use)

The most common case where chunking `astype` would be beneficial is `BF16 -> F16`.
The current behavior for that case is `BF16 -> F32 -> F16`, and chunking `astype` would only save the last `-> F16` part (which is `1x` the size of the input tensor).

Avoiding the BF16 round-trip to F32 would save more memory than chunking `astype`. I think it's still worthwhile to attempt to reduce what is materialized, if we think long term.

Note that `copy=False` is used for `F16` and `F32` in `gguf.quantize()`:

https://github.com/ggml-org/llama.cpp/blob/e8d99dd0b67f2ecc1e45fca8074a3a18c3e036d2/gguf-py/gguf/quants.py#L56-L62

> Even with `--use-temp-file`, the pipeline could still OOM on 100B+ models because `astype` creates a large transient array prior to streaming.

Note that `--use-temp-file` doesn't really use less memory than lazy conversion (unless lazy conversion leaks), and is a legacy option which was kept in case lazy conversion misbehaves (its main use-case is with `--no-lazy`).

There *could* be a way to make something like what you're suggesting more general, though. I think the idea of minimizing what is materialized is interesting. Making it more general might require tracking file ranges (including when stacking MoE experts), and figuring out when they're invalidated, but it would allow avoiding the BF16 round-trip to F32 when possible (simpler solutions for that likely exist, though), and would notably allow using `os.copy_file_range` which could potentially lead to near-instantaneous conversion on Copy-On-Write filesystems.

compilade commented:
I've tested this with <https://huggingface.co/Qwen/Qwen3-30B-A3B-Instruct-2507> (a BF16 MoE model) in the `BF16 -> F16` case, measuring peak memory use with `time -v` (but not the built-in one from `bash`, this is [GNU `time`](https://www.gnu.org/software/time/))

On `master`:

```console
$ $(which time) -v python3 convert_hf_to_gguf.py /path/to/downloaded_model_dir --outfile Qwen3-30B-A3B-Instruct-2507-F16-master.gguf
...
        Maximum resident set size (kbytes): 9165048
...
```

With this PR:

```console
$ $(which time) -v python3 convert_hf_to_gguf.py /path/to/downloaded_model_dir --outfile Qwen3-30B-A3B-Instruct-2507-F16-chunked-astype.gguf
...
        Maximum resident set size (kbytes): 9164960
...
```

**I'm not noticing a significant change**. Both are using a peak of around 9GiB within 88 kilobytes of each other.

@igloo58 How did you measure peak memory usage? Is there another way to reliably measure that than with GNU `time`?

(on MacOS, I think the equivalent is `/usr/bin/time -l`, see <https://stackoverflow.com/a/46874737>)

igloo58 commented:
Hey @compilade thanks for the follow-up!

About your results
The ~±0–12 MB deltas you’re seeing make sense if the streaming path didn’t trigger. This PR only kicks in for pure NumPy astype(...) nodes that reach the lazy writer. In the cases you tested:
	•	Qwen3-30B-A3B (BF16→F16, MoE): tensors are stacked / BF16 round-tripped before write, so there’s no pure astype to stream.
	•	TinyLlama F16→F16 / BF16 / Q8_0: either no cast, or conversions happen before the writer (or via paths that don’t use np.astype). In those paths my change is effectively a no-op, which matches your table.

How I measured (macOS)

```
/usr/bin/time -l python3 convert_hf_to_gguf.py <model_dir> --outfile <out.gguf> 2>&1 | tee logs/run.txt
# I read both “maximum resident set size (kbytes)” and “peak memory footprint (bytes)”.
```

I’ll re-run on Linux with GNU time -v so we’re apples-to-apples with your numbers.

Next steps I’ll take
	1.	Force a pure cast to isolate the path and compare with GNU time -v, e.g.:
	•	Model originally F16 → run with --outtype f32 (exercises np.astype(np.float32)).
	•	Model originally F32 → run with --outtype f16 (exercises np.astype(np.float16)).
I’ll post before/after logs for both.
	2.	Add a debug knob (e.g. GGUF_DEBUG_STREAM=1) that logs when the streaming-cast path is used (dtype→dtype and chunk size). That will let us verify quickly whether a given model actually hits it.
	3.	Follow-up work (separate PR): prototype avoiding the BF16→F32 materialization (either preserve BF16 or stream BF16→F16 directly). That should help the MoE/BF16 case you tested far more than chunking astype.

If you’d like me to use a specific model to exercise the pure-cast path, I’m happy to match that exactly.

compilade commented:
@igloo58 Nice to learn that you did use `time` (even though it's the BSD one in MacOS).

GNU `time` doesn't have a "peak memory footprint" section, though, only "Maximum resident set size". I wonder if there's a way to get something similar in Linux.

> Model originally F16 → run with --outtype f32 (exercises np.astype(np.float32)).

<https://huggingface.co/SpectraSuite/TriLM_3.9B_Unpacked> is a model in F16.

Testing with GNU `time`, and converting that directly to F32:

`master`:

```console
$ $(which time) -v python3 convert_hf_to_gguf.py /srv/models/src/TriLM_3.9B_Unpacked/ --outfile /srv/models/tmp/TriLM_3.9B_Unpacked-F32-master.gguf --outtype f32
...
        Maximum resident set size (kbytes): 5389496
...
```

This PR:

```console
$ $(which time) -v python3 convert_hf_to_gguf.py /srv/models/src/TriLM_3.9B_Unpacked/ --outfile /srv/models/tmp/TriLM_3.9B_Unpacked-F32-chunked-astype.gguf --outtype f32
...
        Maximum resident set size (kbytes): 5499200
...
```

|Model | Temp file | Target type | peak RSS on master (kbytes) | peak RSS on this PR (kbytes) | Delta (kbytes) |
|-|-|-|-|-|-|
| <https://huggingface.co/SpectraSuite/TriLM_3.9B_Unpacked> | No | F32 | 5 389 496 | 5 499 200 | +109 704 |

There is no decrease in memory usage (there's an *increase* of 107MiB, but maybe it's noise). I guess this is because the chunk size of 256 MiB is too big to make the potential reduction from chunking noticeable.

Let's try with a chunk size of 16 MiB (by making `mb` always 16):

```console
$ $(which time) -v python3 convert_hf_to_gguf.py /srv/models/src/TriLM_3.9B_Unpacked/ --outfile /srv/models/tmp/TriLM_3.9B_Unpacked-F32-chunked16MiB-astype.gguf --outtype f32
...
        Maximum resident set size (kbytes): 5382132
...
```

This *is* smaller than `master`, but only by around 7 MiB, which could be considered negligible.

I also checked that the code path for chunked `astype` is reached, and **it is** (at least in the `F16 -> F32` case).

> If you’d like me to use a specific model to exercise the pure-cast path, I’m happy to match that exactly.

If you have another suitable model, I can also match it. I think the best would be one in F16 and with a *huge* vocab so that the token embeddings tensor is enormous. The rest of the model doesn't have to be big. TriLM_3.9B only has 50688 tokens in its vocab.

compilade commented:
@igloo58 Good to see an example use-case with real effects! `bloom-560m` really does have a big vocab.

Regarding the default chunk size, I think it depends on how much RAM the users of the conversion script are expected to have. Before this change (or maybe with #15667), it was `overhead` + `converted biggest tensor` (assuming mmaped input tensor, otherwise also add that), and with this change it is closer to `overhead` + `chunk size`, at least in the cases where it's applicable (which means excluding MoE models which require stacking, because in that case the worst case is unchanged).

For an example of the overheads, with TriLM-3.9B, `overhead` seems to be around `200 MiB` and comes from the size of `torch`, `transformers`, and also the `8.5 MiB` vocab (according to a `memray` flamegraph).

I think `256 MiB` is acceptable, but of course making it lower can be beneficial (unless it's so small the Python loop becomes a bottleneck). If you want to make it `64 MiB` or `32 MiB`, sure.

---

My main concern with this change is that the way it's implemented (wrapping `astype` and `tofile`) means it will break (aka be bypassed) with changes like #12837 in which `tofile` is not used anymore. An other concern is that it's only beneficial in very limited use-cases (although #15667 will likely make this PR have measurable impact in more cases). 

I don't know how to keep both chunked `astype` as proposed here and multi-threaded writes (from #12837).

What would be needed (to make both work together) might be a way to have chunked lazy tensors? i.e. lazy tensors which can be materialized in chunks? Kind of a `to_eager`, but sliced? This *sounds* possible, but I'm not sure how (especially regarding interactions with permutes). I guess it would need to be able to fallback to a full materialization when chunking is not possible (or not obvious).

This is kind of related to the file-range tracking idea at the end of <https://github.com/ggml-org/llama.cpp/pull/15648#issuecomment-3235556780>, because both problems need some kind of mapping between the input tensor(s) and the materialized tensor, but the chunked lazy tensors problem doesn't necessarily require that the data comes from a file. Both problems need to be able to fallback to the normal behavior when they can't handle a particular transformation (and I think the problematic transformations would be the same or very similar for both problems).

igloo58 committed:
gguf-py: stream dtype casts (default 64 MiB), add writer-side path + debug logs

	•	New helper: gguf/stream_cast.py with write_cast(fp, src_arr, dst_dtype, chunk_mb) that writes src.astype(dst) in fixed-size chunks to cap peak RSS.
	•	lazy.py:
	•	tag LazyNumpyTensor.astype() results (_gguf_stream_cast, _gguf_stream_cast_dtype)
	•	tofile() streams via write_cast when the node is a pure dtype cast; otherwise falls back.
	•	env vars: GGUF_CAST_CHUNK_MB (default 64) and GGUF_STREAM_LOG (opt-in diagnostics).
	•	gguf_writer.py: call write_cast directly when the tensor is a tagged pure cast. This keeps the benefit even if future changes bypass tofile() / use multi-threaded writes.
	•	Alignment: preserve data_alignment by padding before/after writes.
	•	Repro notes: Ubuntu 24.04 / Python 3.12 / NumPy 2.1; bloom-560m FP16→F32 conversion shows peak RSS reductions when chunking (e.g., 256→64→32→16 MiB) with small runtime trade-offs at smaller chunks. macOS run logs confirm [gguf-stream] activation as well.
	•	Scope & limitations: only pure dtype casts; MoE stacking / complex transforms fall back.
	•	Future work (separate RFC/PR): “chunked lazy tensors” and file-range tracking compatible with multi-threaded writes.
